### PR TITLE
fix($location): do not decode forward slashes in the path in HTML5 mode

### LINKED ISSUE #16312: AngularJS unescapes matrix parameters
**I'm submitting a ...**
- [x] bug report
- [ ] feature request
- [ ] other (Please do not submit support requests here (see above))

**Current behavior:**
AngularJS un-escapes matrix parameters, preventing matrix parameters with slashes from being used in dual-router AngularJS / Angular applications.

**Expected / new behavior:**
AngularJS should not un-escape matrix parameters. 

**Minimal reproduction of the problem with instructions:**
In a hybrid setup, add a link like: <a [routerLink]="['/a/ng2', {path: '/some/path'}]">ANGULAR A</a>
While Angular properly renders this link as "http://localhost:4200/a/ng2;path=%2Fsome%2Fpath", AngularJS rewrites it as http://localhost:4200/a/ng2;path=/some/path during navigation.  Depending on the setup, this causes the application to display either a missing route mapping or a redirect loop.

**AngularJS version:** 1.5+

**Browser:** all

### PR DESCRIPTION
**Pull request by petebacondarwin**
Closes #16312

**What kind of change does this PR introduce? (Bug fix, feature, docs update, ...)**

bug fix

**What is the current behavior? (You can also link to an open issue here)**

See #16312

**What is the new behavior (if this is a feature change)?**

**Does this PR introduce a breaking change?**

It depends what people expected of the location escaping...
I think what we had before was buggy. So no.

**Please check if the PR fulfills these requirements**
- [x] The commit message follows our guidelines: https://github.com/angular/angular.js/blob/master/CONTRIBUTING.md#commit-message-format
- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

**Other information**:

### CODE REVIEW

**Comment by petebacondarwin** [comment]
Be aware that this does not change the behaviour of hash-bang (non-HTML5 mode) paths. This is because slashes do not need to be escaped once they are in the hash section of the URI and we have tests that expect encoded slashes to be decoded into the path when parsed by `$location`. Thus changing that behaviour would most likely constitute a breaking change.

The use case in #16312 is very unlikely to use hashbang mode since it is a hybrid AngularJS/Angular app that would run on browsers that support HTML5 mode.

File: src/ng/location.js [hunk 1]
```diff
@@ -16,7 +16,23 @@ function encodePath(path) {
...
+    // decode forward slashes to prevent them from being double encoded
+    segments[i] = encodeUriSegment(segments[i].replace(/%2F/g, '/'));
+  }
+
+  return segments.join('/');
+}
+
+function decodePath(path, html5Mode) {
+  var segments = path.split('/'),
+  i = segments.length;
```
Comment (gkalpak) [inline src/ng/location.js]: No indentation? 😞

**Review by gkalpak — APPROVED** [review]
I am very sure someone's app just broke :smiley:

File: src/ng/location.js (hunk 1 above)
Comment (petebacondarwin) [reply src/ng/location.js]: :scream:

**Comment by veloek** [comment]
@gkalpak You're so right! The change from decodeURIComponent to decodeURI on line 425 broke our URL matcher type in ui-router since decodeURI doesn't decode %2F to / in the path (we don't use html5-mode).

I would like to see this one line reverted, but I'm not sure what the implications are...

Please let me know if you need any more info from me.

**Comment by veloek** [comment]
@frederikprijck Yup, that's what I did :)

**Comment by petebacondarwin** [comment]
@veloek - can you provide a reproduction of the regression for us?

**Comment by veloek** [comment]
@petebacondarwin I'll see if I get the time :)

I guess the locationPrototype.url function is used in both html5mode and "hash-mode", right?

**Comment by petebacondarwin** [comment]
Hmm, I see. We probably need to shift some of that logic into the different implementations: `LocationHtml5Url` and `LocationHashbangUrl`.

**Comment by petebacondarwin** [comment]
Here is a fix for the regression: https://github.com/angular/angular.js/pull/16350

**Pull request merged**